From 78e358bcaa9a73c8ac86264236c12e75718eb55a Mon Sep 17 00:00:00 2001
From: Daniel Goehring <dgoehrin@os.amperecomputing.com>
Date: Tue, 6 Jun 2023 14:44:13 -0600
Subject: [PATCH 16/34] target/aarch64: add missing aarch64_poll() calls

The code in aarch64_step() does not check the core status before
returning. Add call to aarch64_poll() before returning from
aarch64_step() to ensure the event TARGET_EVENT_HALTED is called.

Also add the aarch64_poll() call to aarch64_init_debug_access()
to update target state information.

This is necessary with the poller update introduced in commit
95603fae18f8 ("openocd: revert workarounds for 'expr' syntax change")

Signed-off-by: Daniel Goehring <dgoehrin@os.amperecomputing.com>
Change-Id: I6e91f1b6bc1f0d16e6f0eb76fc67d20111e3afd2
---
 src/target/aarch64.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/src/target/aarch64.c b/src/target/aarch64.c
index 1ca93f75a..5c336df75 100644
--- a/src/target/aarch64.c
+++ b/src/target/aarch64.c
@@ -250,6 +250,10 @@ static int aarch64_init_debug_access(struct target *target)
 
 	/* Resync breakpoint registers */
 
+	/* Since this is likely called from init or reset, update target state information */
+	if (retval == ERROR_OK)
+		retval = aarch64_poll(target);
+
 	return retval;
 }
 
@@ -1140,6 +1144,7 @@ static int aarch64_step(struct target *target, int current, target_addr_t addres
 	struct aarch64_common *aarch64 = target_to_aarch64(target);
 	struct target_list *head;
 	int saved_retval = ERROR_OK;
+	int poll_retval;
 	int retval;
 	uint32_t edecr;
 
@@ -1249,6 +1254,9 @@ static int aarch64_step(struct target *target, int current, target_addr_t addres
 	if (retval == ERROR_TARGET_TIMEOUT)
 		saved_retval = aarch64_halt_one(target, HALT_SYNC);
 
+	/* Update target state information */
+	poll_retval = aarch64_poll(target);
+
 	/* restore EDECR */
 	retval = mem_ap_write_atomic_u32(armv8->debug_ap,
 			armv8->debug_base + CPUV8_DBG_EDECR, edecr);
@@ -1265,6 +1273,9 @@ static int aarch64_step(struct target *target, int current, target_addr_t addres
 	if (saved_retval != ERROR_OK)
 		return saved_retval;
 
+	if (poll_retval != ERROR_OK)
+		return poll_retval;
+
 	return ERROR_OK;
 }
 
-- 
2.25.1

